README.md
![Build Status](https://img.shields.io/docker/pulls/n0r1skcom/border-controller.svg) ![Build Status](https://img.shields.io/docker/automated/n0r1skcom/border-controller.svg)  ![Build Status](https://img.shields.io/docker/build/n0r1skcom/border-controller.svg) [![Build Status](https://travis-ci.org/n0r1sk/border-controller.svg?branch=edge)](https://travis-ci.org/n0r1sk/border-controller)

[![Anchore Image Overview](https://anchore.io/service/badges/image/68f00f08cde7b43f90ad3ce9a3a48bf282e649e1bd6854df47e7875f9d1f5882)](https://anchore.io/image/dockerhub/n0r1skcom%2Fborder-controller%3A1.0.1)

# border-controller
This is a Nginx based ingress border controller with automatic configuration reload based on Docker swarm DNS service discovery for on-premise, but not limited to, use.

# Why?
The current problem in the Docker swarm infrastructure is, that the swarm mesh network does currently not support sticky connections. We know, that for example [Traefik](https://docs.traefik.io/) exists, which covers many if not all problems mentioned here. But currently, Traefik does not support TCP load balancing and you will have services, which are not using the HTTP protocol. Furthermore, there will be setups, where you won't let Traefik communicate with the Docker swarm manager.

Please don't get this wrong, but if Traefik has an error, Traefik can remove all your Docker stacks as it is communicating with the Docker swarm manager. We know, that it is only reading information from the Docker swarm service, but it can also send commands there if someone implements it. In our personal opinion, the Docker swarm DNS based service discovery is very useful to retrieve the backend ip address container information. This is, what our binary does.

Another thing is, that Nginx has a lot of configuration possibilities which you might like to have. Therefore this project is based on the Golang text template system. You can write whatever Nginx config you like and replace the backend information with data from the backends.

# Conclusion
This project is far away from being perfect, nor does it reflect perfect written code. It just works and maybe there is someone out there who find it useful. For us it is also nice to have an interface to the [PowerDNS API](https://www.powerdns.com) because an automatic registration of the ip address with the DNS server is extram helpful in a large environment where you have to manage more than two or three services. This project is limited to Docker swarm services!

# Configuration

Here is an example configuration.

## border-controller
```
debug: true
general:
  check_intervall: 10
  resources:
    testcontexta:
      context: /context/a
      port: 8080
      task_dns: tasks.testa.app
    testcontextb:
      context: /context/b
      port: 9090
      task_dns: tasks.testb.app
pdns:
  api_url: https://your.pdns/api/v1/servers/localhost/zones
  api_key: yourapikey
  ip_address: 1.1.1.1
  domain_prefix: funny
  domain_zone: domain.com

```
### debug
If parameter is not provided, default value is set to false.

### check_intervall
If parameter is not provided, default value set to 30 secons. Time duration is in seconds.

### pdns
This configuration section is completely optional.

## nginx.conf
In the ```nginx.conf``` you can do all configuration which is supported by Nginx. There are no restrictions.

```
worker_processes auto;

events {}

http {

    upstream {{.testcontexta.Upstream}} {
       hash $remote_addr;
       {{range $index, $entry := .testcontexta.Servers}} server {{$entry.Server}}:{{$entry.Port}};
       {{end}}
    }

    upstream {{.testcontextb.Upstream}} {
       hash $remote_addr;
       {{range $index, $entry := .testcontexta.Servers}} server {{$entry.Server}}:{{$entry.Port}};
       {{end}}
    }

    server {
        listen 80;
    }
}

```

# Changelog
You can find the changelog information [here](CHANGELOG.md).

# Version
1.2


CHANGELOG.md
# v1.2
## Features
* Add PowerDNS API support

## Fixes
* Ignore missing binary
* Restart or start even if services are not resolvable by Docker swarm DNS
* Reread configuration every time interval
* Modify the image build to include tools like ping
* Restart or start when the Nginx process is a zombie (configuration error)

## General
* Colorize output
* Cleanup logging code
* Cleanup documentation
* Add badges for Docker hub information


powerdns.go
/*
Copyright 2017 Mario Kleinsasser and Bernhard Rausch

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"bytes"
	"encoding/json"
	"io/ioutil"
	"net/http"

	log "github.com/sirupsen/logrus"
)

type Recordsets struct {
	Rrsets []*Recordset `json:"rrsets"`
}

type Recordset struct {
	Name       string   `json:"name"`
	Type       string   `json:"type"`
	Ttl        int      `json:"ttl"`
	Changetype string   `json:"changetype"`
	Records    []Record `json:"records"`
}

type Record struct {
	Content  string `json:"content"`
	Disabled bool   `json:"disabled"`
}

func recordsetstojson(data Recordsets) []byte {

	// set time to live to 10 if not set
	for _, e := range data.Rrsets {
		p := &e.Ttl
		if *p == 0 {
			*p = 10
		}
	}

	b, err := json.Marshal(data)
	if err != nil {

	}

	log.Debug(string(b[:]))
	return b

}

func recordsetsreplace(api_url string, api_key string, domain_zone string, data Recordsets) {
	s := recordsetstojson(data)

	log.Debug(string(s[:]))

	req, err := http.NewRequest("PATCH", api_url+"/"+domain_zone+".", bytes.NewBuffer(s))
	req.Header.Set("X-API-Key", api_key)

	if err != nil {
		log.Warn(err)
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		log.Warn(err)
		return
	}
	defer resp.Body.Close()

	body, _ := ioutil.ReadAll(resp.Body)
	log.Debug("response Body:", string(body))
	log.Info("PDNS Update response code: " + resp.Status)
}

func updatepdns(config T) {
	var rrset Recordsets
	var rs Recordset
	var r Record

	r.Content = config.Pdns.Ip_address
	r.Disabled = false

	rs.Name = config.Pdns.Domain_prefix + "." + config.Pdns.Domain_zone + "."
	rs.Changetype = "REPLACE"
	rs.Type = "A"
	rs.Records = append(rs.Records, r)

	rrset.Rrsets = append(rrset.Rrsets, &rs)

	recordsetsreplace(config.Pdns.Api_url, config.Pdns.Api_key, config.Pdns.Domain_zone, rrset)
}


configreader.go
/*
Copyright 2017 Mario Kleinsasser and Bernhard Rausch

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	//"fmt"
	"io/ioutil"
	"log"

	"gopkg.in/yaml.v2"
)

type T struct {
	General struct {
		Swarm struct {
			Docker_hosts           []string
			Docker_host_dns_domain string
			Ingress_service_name   string
			Docker_controller      struct {
				Api_key      string
				Exposed_port string
			}
		}
	}
}

func ReadConfigfile() (config T, ok bool) {
	cfgdata, err := ioutil.ReadFile("/config/border-controller.yml")

	if err != nil {
		log.Panic("Cannot open config file from /config/border-controller.yml")
	}

	t := T{}

	err = yaml.Unmarshal([]byte(cfgdata), &t)
	if err != nil {
		log.Panic("Cannot map yml config file to interface, possible syntax error")
		log.Panic(err)
	}

	return t, true
}


config-reader.go
/*
Copyright 2017 Mario Kleinsasser and Bernhard Rausch

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"io/ioutil"
	"log"

	"gopkg.in/yaml.v2"
)

type Backendcfg struct {
	Upstream      string
	Context       string
	Servers       []Backend
	Port          string
	Task_dns      string
	Domain_prefix string
	Domain_zone   string
}

type T struct {
	Debug   bool
	General struct {
		Check_intervall int64
		Domain_prefix   string
		Domain_zone     string
		Resources       map[string]*Backendcfg
	}
	Pdns struct {
		Api_url       string
		Api_key       string
		Ip_address    string
		Domain_prefix string
		Domain_zone   string
	}
}

func ReadConfigfile() (ok bool, config T) {
	cfgdata, err := ioutil.ReadFile("/config/border-controller.yml")

	if err != nil {
		log.Panic("Cannot open config file from /config/border-controller.yml")
	}

	t := T{}

	err = yaml.Unmarshal([]byte(cfgdata), &t)
	if err != nil {
		log.Panic("Cannot map yml config file to interface, possible syntax error")
		log.Panic(err)
	}

	return true, t
}


ingress-controller.go
/*
Copyright 2017 Mario Kleinsasser and Bernhard Rausch

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"bytes"
	"crypto/md5"
	"fmt"
	"io/ioutil"
	"log"
	"net"
	"net/http"
	"os"
	"os/exec"
	"sort"
	"text/template"
	"time"
)

var issd string
var issp string
var issthp string
var mainloop bool

type Backend struct {
	Hostname string
	Port     string
}

func isprocessrunning() (running bool) {
	// check if traefik is healthy
	var run bool
	run = true

	resp, err := http.Get("http://localhost:" + issthp + "/health")

	if err != nil {
		log.Print("Error getting traefik status. Trying to start traefik. NIL!")
		log.Print(err)
		return false
	}

	if resp.StatusCode != 200 {
		log.Print("Error getting traefik status. Trying to start traefik.")
		log.Print(string(resp.StatusCode))
		return false
	}

	log.Print("Traefik is running. All OK!")
	return run
}

func startprocess() {
	log.Print("Start Process!")
	cmd := exec.Command("/data/traefik", "-c", "/config/traefik.toml")
	err := cmd.Start()
	if err != nil {
		log.Fatal(err)
		mainloop = false
	}

}

func checkconfig(backends []Backend) {

	//  open template
	t, err := template.ParseFiles("/config/ingress-controller-config.tpl")
	if err != nil {
		log.Print(err)
		return
	}

	// process template
	var tpl bytes.Buffer
	err = t.Execute(&tpl, backends)
	if err != nil {
		log.Print(err)
		return
	}

	// create md5 of result
	md5tpl := fmt.Sprintf("%x", md5.Sum([]byte(tpl.String())))
	log.Print("MD5 of TPL: " + md5tpl)
	log.Print("TPL: " + tpl.String())

	// open existing config, read it to memory
	exconf, errexconf := ioutil.ReadFile("/config/traefik.toml")
	if errexconf != nil {
		log.Print("Cannot read existing conf!")
		log.Print(errexconf)
	}

	md5exconf := fmt.Sprintf("%x", md5.Sum(exconf))
	log.Print("MD5 of EXCONF: " + md5exconf)

	// comapre md5 and write config if needed
	if md5tpl == md5exconf {
		log.Print("MD5 sums equal! Nothing to do.")
		return
	}

	log.Print("MD5 sums different writing new conf!")

	// overwrite existing conf
	err = ioutil.WriteFile("/config/traefik.toml", []byte(tpl.String()), 0644)
	if err != nil {
		log.Print("Cannot write config file.")
		log.Print(err)
		mainloop = false
	}

}

func querydns() (bends []Backend) {

	// we dont kill the mainloop if dns not working because it could come online later
	var data []string
	var backends []Backend
	var b Backend

	// lookup dns name
	be, errr := net.LookupIP(issd)

	if errr != nil {
		log.Print(errr)
	}

	// save to simple string slice
	for _, ip := range be {
		data = append(data, ip.String())
	}

	sort.Strings(data)

	// convert it to struct
	for _, ip := range data {
		b.Hostname = ip
		b.Port = issp
		backends = append(backends, b)
	}

	return backends
}

func checkenvironment() {

	var errb bool

	// The following environments must be present
	issd, errb = os.LookupEnv("INGRESS_STACK_SERVICE_DNS")
	if (!errb) || (issd == "") {
		log.Panic("No environment variable INGRESS_STACK_SERVICE_DNS or empty value!")
	}

	issp, errb = os.LookupEnv("INGRESS_STACK_SERVICE_PORT")
	if (!errb) || (issp == "") {
		log.Panic("No environment variable INGRESS_STACK_SERVICE_PORT or empty value!")
	}

	issthp, errb = os.LookupEnv("INGRESS_STACK_TRAEFIKHEALTH_PORT")
	if (!errb) || (issp == "") {
		log.Panic("No environment variable INGRESS_STACK_TRAEFIKHEALTH_PORT or empty value!")
	}

}

func main() {

	checkenvironment()

	// now checkconfig, this will loop forever
	mainloop = true
	for mainloop == true {

		backends := querydns()

		checkconfig(backends)

		if !isprocessrunning() {
			startprocess()
		}

		time.Sleep(10 * time.Second)
	}
}


border-controller.go
/*
Copyright 2017 Mario Kleinsasser and Bernhard Rausch

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"bufio"
	"bytes"
	"crypto/md5"
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"os/exec"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"text/template"
	"time"

	log "github.com/sirupsen/logrus"
	. "github.com/logrusorgru/aurora"
)

var mainloop bool
var ctrlcmd *exec.Cmd

type Message struct {
	Acode   int64
	Astring string
	Aslice  []string
}

type Backend struct {
	Server string
	Port   string
}

func isprocessrunningps(processname string) (running bool) {

	// get all folders from proc filesystem
	running = false

	files, _ := ioutil.ReadDir("/proc")
	for _, f := range files {

		// check if folder is a integer (process number)
		if _, err := strconv.Atoi(f.Name()); err == nil {
			// open status file of process
			f, err := os.Open("/proc/" + f.Name() + "/status")
			if err != nil {
				log.Info(err)
				return running
			}

			// read status line by line
			scanner := bufio.NewScanner(f)

			// check if process name in status of process
			var process bool
			process = false

			for scanner.Scan() {


				re := regexp.MustCompile("^Name:.*" + processname + ".*")
				match := re.MatchString(scanner.Text())

				if match == true {
					running = true
					process = true
				}

				if process == true {
					re := regexp.MustCompile("^State:.*Z.*")
					match := re.MatchString(scanner.Text())
					if match == true {
						log.Warn(Sprintf(Bold(Magenta("Error ZOMBIE process! Config error?"))))
						// The process seems to be dead, call wait on it
						// to bury the child process from the parent
						err := ctrlcmd.Wait()
						if err != nil{
							log.Warn(Sprintf(Magenta("%s"),err.Error()))
						}
					}
					running = true
				}

			}
			if running == true{
				return running
			}

		}

	}

	return running

}

func startprocess() {
	log.Info("Start Process!")
	cmd := exec.Command("nginx", "-g", "daemon off;")
	err := cmd.Start()

	if err != nil {
		log.Warn(Sprintf(Cyan("%s: "),err.Error()))
	}
	ctrlcmd = cmd

	// just give the process some time to start
	time.Sleep(time.Duration(250) * time.Millisecond)
	ok := isprocessrunningps("nginx")
	if ok == true{
		log.Info(Green("Process started"))
	}
}

func reloadprocess() {
	log.Info("Reloading Process!")
	cmd := exec.Command("nginx", "-s", "reload")
	err := cmd.Start()
	if err != nil {
		log.Warn(Sprintf(Cyan("%s: "),err.Error()))
	}
	cmd.Wait()
	isprocessrunningps("nginx")
}

func writeconfig(data interface{}) (changed bool) {

	//  open template
	t, err := template.ParseFiles("/config/border-controller-config.tpl")
	if err != nil {
		log.Info(err)
		return false
	}

	// process template
	var tpl bytes.Buffer
	err = t.Execute(&tpl, data)
	if err != nil {
		log.Info(err)
		return false
	}

	// create md5 of result
	md5tpl := fmt.Sprintf("%x", md5.Sum([]byte(tpl.String())))
	log.Debug("MD5 of TPL: " + md5tpl)
	log.Debug("TPL: " + tpl.String())

	// open existing config, read it to memory
	exconf, errexconf := ioutil.ReadFile("/etc/nginx/nginx.conf")
	if errexconf != nil {
		log.Warn("Cannot read existing conf!")
		log.Warn(errexconf)
	}

	md5exconf := fmt.Sprintf("%x", md5.Sum(exconf))
	log.Debug("MD5 of EXCONF: " + md5exconf)

	// comapre md5 and write config if needed
	if md5tpl == md5exconf {
		log.Info(Green("MD5 sums equal! Nothing to do."))
		return false
	}

	log.Info(Brown("MD5 sums different writing new conf!"))

	// overwrite existing conf
	err = ioutil.WriteFile("/etc/nginx/nginx.conf", []byte(tpl.String()), 0644)
	if err != nil {
		log.Warn("Cannot write config file.")
		log.Warn(err)
		mainloop = false
	}

	return true

}

func getstacktaskdns(task_dns string) (addrs []string, err error) {

	// resolve given service names
	servicerecords, err := net.LookupHost(task_dns)
	sort.Strings(servicerecords)

	if err != nil {
		return nil, err
	}

	log.Debug("TASK_DNS: " + task_dns + " ENTRIES: " + strings.Join(servicerecords, " "))

	return servicerecords, nil

}

func refreshconfigstruct(config T) (ok bool, err error) {
	// get information on services and context configuration information

	ok = true

	log.Debug("Config Struct: " + fmt.Sprintf("%+v", config))
	for k, v := range config.General.Resources {
		v.Servers = nil

		// there must be a dns name given and we have to get the backend ips
		servicerecords, err := getstacktaskdns(v.Task_dns)

		if err != nil {
			log.Warn("Cannot get DNS records for config entry: " + k)
			ok = false
		}

		log.Info(Sprintf(Cyan("%s: "),k) + fmt.Sprintf("%+v", servicerecords))
		for _, s := range servicerecords {
			var b Backend
			b.Server = s
			b.Port = v.Port
			v.Servers = append(v.Servers, b)
		}

		// set the key to upstream
		v.Upstream = k

		// add values of DNS to config struct
		if v.Domain_zone == "" {
			if config.General.Domain_zone != "" {
				v.Domain_zone = config.General.Domain_zone
			} else if config.Pdns.Domain_zone != "" {
				v.Domain_zone = config.Pdns.Domain_zone
			} else {
				// leave empty
			}
		}

		if v.Domain_prefix == "" {
			if config.General.Domain_prefix != "" {
				v.Domain_prefix = config.General.Domain_prefix
			} else if config.Pdns.Domain_prefix != "" {
				v.Domain_prefix = config.Pdns.Domain_prefix
			} else {
				// leave empty
			}
		}

		// fill the dns values
		log.Debug("Config Struct: " + fmt.Sprintf("%+v", v))

	}

	return ok, nil

}

func main() {

	// configure logrus logger
	customFormatter := new(log.TextFormatter)
	customFormatter.TimestampFormat = "2006-01-02 15:04:05"
	customFormatter.FullTimestamp = true
	customFormatter.ForceColors = true
	log.SetFormatter(customFormatter)
	log.SetOutput(os.Stdout)

	ok, config := ReadConfigfile()
	if !ok {
		log.Warn(Sprintf(Red("Error during config parsing, yet continuing!")))
	}

	// get debug flag from config
	if config.Debug == true {
		log.SetLevel(log.DebugLevel)
	}

	// set check intervall from config
	checkintervall := config.General.Check_intervall
	if checkintervall == 0 {
		checkintervall = 30
	}

	// check if pdns configuration is enabled
	if config.Pdns.Api_key != "" {
		updatepdns(config)
	}

	// now checkconfig, this will loop forever
	mainloop = true
	var changed bool = false

	for mainloop == true {
		// reread config file
		ok, config := ReadConfigfile()
		if !ok {
			log.Warn(Sprintf(Red("Error during config parsing, yet continuing!")))
		}

		// refresh config struct
		ok, err := refreshconfigstruct(config)
		if err != nil {
			// on error during refresh (DNS) sleep and continue
			time.Sleep(time.Duration(checkintervall) * time.Second)
		}

		if ok == false {
			log.Warn(Sprintf(Red("Error during refresh of config, yet starting nginx!")))
		}

		// process config
		changed = writeconfig(config.General.Resources)

		if changed == true {
			if isprocessrunningps("nginx") {
				reloadprocess()
			} else {
				startprocess()
			}
		} else {
			if !isprocessrunningps("nginx") {
				startprocess()
			}
		}

		time.Sleep(time.Duration(checkintervall) * time.Second)
	}

}
